### DISCUSSION TITLE
Reproduced ChatGPT2 (124M): pretrained model + SFT + RLHF !!!

### DISCUSSION THREAD

**Discussion by jacoboromerodiaz** [karpathy/build-nanogpt#113] [Show and tell]
Hi everyone,
<p>I just finished a project where I built GPT-2 following Karpathy's tutorial and took it all the way to a full instruct model: pretraining, <strong>SFT</strong>, and <strong>RLHF</strong>. My goal wasn't to beat any benchmark or other comments in the discussion. It was to understand every part of LLM <strong>end-to-end</strong> development.</p>
<p>A pretrained base just predicts the next token on raw text, it doesn't follow instructions, doesn't know what an assistant turn looks like, and has no notion of a "preferred" output. So, I continued the pipeline by implementing:</p>
<p>🔧 <strong>Stage 1: Supervised Finetuning</strong> </p>
<ul>
<li>This shifts the base from raw next-token prediction → <strong>assistant</strong> format.</li>
<li>Training on Alpaca-cleaned + Dolly 15K, formatted as user/assistant turns. </li>
<li>Only the assistant response tokens contribute to the loss; user prompts are masked with -100 so the model doesn't memorize queries. </li>
</ul>
<p>🔧 <strong>Stage 2: Reinforcement Learning from Human Feedback</strong> </p>
<ul>
<li>This part was more complex to implement, where I took some lessons by <a href="https://www.youtube.com/watch?v=jQPiH-KB4B0&#x26;list=PLL1tdVxB1CpVpEtMHxwuR4uI4Lxjw00_y" class="external-link" target="_blank" rel="noopener nofollow" aria-label="https://www.youtube.com/watch?v=jQPiH-KB4B0&#x26;list=PLL1tdVxB1CpVpEtMHxwuR4uI4Lxjw00_y" data-tooltip-position="top">Nathan Lambert on RLHF</a>.</li>
<li>Used GRPO instead of vanilla PPO. The value network is dropped and advantages are computed group-relative across completions per prompt, with a KL penalty against the frozen SFT reference keeping the policy from drifting. </li>
<li>Training a reward model from scratch was skipped by using OpenAssistant's pretrained DeBERTa-v3 as the reward signal.</li>
</ul>
<p>These stages are <strong>explained</strong> in detail in my <a href="https://github.com/jacoboromerodiaz/gpt-2" class="external-link" target="_blank" rel="noopener nofollow" aria-label="https://github.com/jacoboromerodiaz/gpt-2" data-tooltip-position="top">repo</a>. </p>
<p>🛠️ <strong>Other extensions over the baseline</strong></p>
<ul>
<li>BPE tokenizer built from scratch.</li>
<li>Seasonality-aware data sampling to reduce distribution skew during pretraining.</li>
<li>Model unwrapping before HellaSwag eval to keep <code>torch.compile</code> active and cut eval time</li>
<li>Checkpoint resume</li>
</ul>
<p>💰 <strong>TOTAL COST: ~$10</strong> (RTX A5000)</p>
<ul>
<li>Pretraining: ~$7.50 (≈37h, 1× 24GB)</li>
<li>SFT: ~$0.22 (&#x3C;1h, 1× 8GB)</li>
<li>RLHF: ~$1.74 (≈8h, 1× 24GB)</li>
</ul>
<p>📈 <strong>HellaSwag across stages</strong></p>

Stage | HellaSwag
-- | --
Pretrained | 0.3066
SFT | 0.3080
RLHF | 0.3174

<p>Qualitative outputs are in <code>finetune/playground.ipynb</code>.</p>
<ul>
<li>Pretrained model:</li>
</ul>
<pre><code class="language-text">> How do I make an orange juice? 
If you want to make an orange juice, you can make an orange juice. A juice is made by mixing a juice with water and a sugar. If you want to make an orange juice, you can make an orange juice. How do I make a lemonade? How to Make a Lemonade. Lemonade is a liquid that can be used to make a lemonade. ...
</code></pre>
<ul>
<li>SFT model</li>
</ul>
<pre><code class="language-text">&#x3C;|im_start|>user 
How do I make an orange juice?&#x3C;|im_end|> 
&#x3C;|im_start|>assistant 
To make an orange juice, you need to add some water, a blender, some sugar, and a glass of water. You can also add some lemon juice and some honey.&#x3C;|im_end|>
</code></pre>
<ul>
<li>RLHF model</li>
</ul>
<pre><code class="language-text">&#x3C;|im_start|>user 
How do I make an orange juice?&#x3C;|im_end|> 
&#x3C;|im_start|>assistant 
Making an orange juice is easy! Here's how to make it: 
1. Start by making a clean and well-drained glass of water. 
2. Next, add a small amount of orange juice to the glass. 
3. Add a few drops of lemon juice to the water. 
4. Once the juice is well absorbed, add a few drops of orange juice to the glass. 
5. Allow the juice to stand for a few minutes before adding more orange juice to the glass. 
6. Once the juice is fully absorbed, add a few drops of orange juice to the glass. 
7. Once the juice is fully absorbed, place the glass in a colander and wait for it to cool. 
8. Once the juice is cool, it can be added to the glass. 
9. Once the juice is fully absorbed, it can be added to the glass to create a delicious orange juice.&#x3C;|im_end|>
</code></pre>
<p>Watching the model progress from raw next-token prediction to assistant format, to preference-aligned outputs was super interesting. If anyone needs help to reproduce my models or discuss any potential improvements/fixes, feel free to reach out!</p>
<p>Repo: <a href="https://github.com/jacoboromerodiaz/gpt-2/tree/main" class="external-link" target="_blank" rel="noopener nofollow">https://github.com/jacoboromerodiaz/gpt-2</a></p>
